### PR TITLE
Add two new flags to opt-in to GUID matching

### PR DESCRIPTION
**Pull request by hughsie**
Add two new flags to control if GUID matching is enabled for either coldplug or
replug. It is far too easy to forget to set FWUPD_DEVICE_FLAG_NO_GUID_MATCHING
for new plugins, and without it it all works really well *until* you have two
devices of the same type installed at the same time and then one 'disappears'
for hard to explain reasons. Typically we only need it for replug anyway!

Add two private device flags to explicitly opt-in to this rarely-required
behaviour, with the default to just use the physical and logical IDs. Also
document the update behavior for each plugin to explain why the flags is used.

This allows you to have two identical Unifying plugged in without one of them
being hidden from the user, at the same time allowing a HIDRAW<->USB transition
when going to and from bootloader and runtime modes.

This removes the workaround added in 99eb3f06b651c34a01acef20d1fde96e4996fffe.

Fixes https://github.com/fwupd/fwupd/issues/2915

### CODE REVIEW

**Comment by hughsie** [comment]
If you're testing this you also need the commit from https://github.com/fwupd/fwupd/pull/2934

File: libfwupdplugin/fu-device.c [hunk 1]
```diff
@@ -2906,6 +2916,16 @@ fu_device_detach (FuDevice *self, GError **error)
...
 		return TRUE;
 
+	/* fixup old plugins */
+	if (priv->has_counterpart &&
+	    !fu_device_has_internal_flag (self, FU_DEVICE_INTERNAL_FLAG_REPLUG_MATCH_GUID)) {
+		g_warning ("device %s [%s] did not set FU_DEVICE_INTERNAL_FLAG_REPLUG_MATCH_GUID "
+			   "so setting it for backwards compat -- this will stop soon!",
+			   fu_device_get_name (self),
+			   fu_device_get_id (self));
+		fu_device_add_internal_flag (self, FU_DEVICE_INTERNAL_FLAG_REPLUG_MATCH_GUID);
```
Comment (superm1) [inline libfwupdplugin/fu-device.c:2926]: the number of plugins that use counterpart is a fixed number, why not just fix those immediately and set an error on detach?

Comment (hughsie) [reply libfwupdplugin/fu-device.c:2926]: I think I fixed them all. There are actually plugins using counterpart for other reasons, e.g. the vli plugin uses it to find the USB-2 device that corresponds to the USB-3 device. Maybe dropping this chunk makes sense.

File: plugins/dell-dock/README.md [hunk 2]
```diff
@@ -48,6 +48,12 @@ These devices use several different generation schemes, e.g.
  * MST Hub: `MST-panamera-vmm5331-259`
  * Thunderbolt Controller: `TBT-00d4b070`
 
+Update Behavior
+---------------
+
+For devices implementing "passive flow" the update is deployed at runtime and
+then actually written the next time the dock is attached.
```
Comment (superm1) [inline plugins/dell-dock/README.md:55]: Come to think of it, NOT using passive was only pre-production devices.  I wonder if it's worth just ripping out the code for non-passive update now.

**Comment by superm1** [comment]
> This removes the workaround added in 99eb3f0.

If doing that - it sounds to me like the devices it helped need `FU_DEVICE_INTERNAL_FLAG_COLDPLUG_MATCH_GUID` set don't they?

**Comment by superm1** [comment]
> If doing that - it sounds to me like the devices it helped need FU_DEVICE_INTERNAL_FLAG_COLDPLUG_MATCH_GUID set don't they?

The case that needed it was `uefi_capsule` + `nvme` both using the same GUID.

File: plugins/dell-dock/README.md (hunk 2 above)
Comment (hughsie) [reply plugins/dell-dock/README.md:55]: I think ripping out code for immediate update (on the assumption they're not what was ever sold in shops) is the best plan. Do you mind doing that and updating the README please? You can probably describe the delayed write thing better than me.

**Comment by hughsie** [comment]
> The case that needed it was uefi_capsule + nvme both using the same GUID.

So, if we actually do this explicit opt-in case I don't actually think anything needs the FU_DEVICE_INTERNAL_FLAG_COLDPLUG_MATCH_GUID behaviour. Certainly not NVMe or UEFI-capsule anymore, and I think the COLDPLUG_MATCH_GUID was really a _load-bearing-bug_ until I worked out the problem in https://github.com/fwupd/fwupd/commit/253e6e4f17fd70661a1f5a7e0cb48503d11cb12e#diff-ceb6643e2c2b41f6ff57f2f230d996792efff6e0da680581d6d445a38f445dceR6000

File: plugins/dell-dock/README.md (hunk 2 above)
Comment (superm1) [reply plugins/dell-dock/README.md:55]: The main thing needed for passive flow was a new enough EC and a new enough hub.  The earliest hub in public release was 01.47 (https://fwupd.org/lvfs/devices/com.dell.wd198ceeeffd.firmware) and minimum EC was 00.00.00.23 (torn out here https://github.com/fwupd/fwupd/commit/ead0deaa255f08db0d777deb97f650d10d545eb5#diff-2d320a289c64c6066499373624e1d9b944d4e13b0eccd3068dad84fda2afd804).

I'll finish tearing out the rest then.

File: plugins/thunderbolt/README.md [hunk 3]
```diff
@@ -44,6 +44,12 @@ The retimer index is oriented around the physical connection within
 the machine.  It is important as multiple controllers may otherwise
 identify identically.
 
+Update Behavior
+---------------
+
+The firmware is written to the device at runtime. Once complete the controller
+may reboot which may cause all connected USB devices to be reenumerated.
```
Comment (superm1) [inline plugins/thunderbolt/README.md:51]: This probably needs some elaboration since flashing WD19 and config option change it.

**Review by superm1 — APPROVED** [review]
just that documentation note for thunderbolt and then LGTM

File: plugins/thunderbolt/README.md (hunk 3 above)
Comment (hughsie) [reply plugins/thunderbolt/README.md:51]: Got anything more please? You understand this much better than me.

**Comment by hughsie** [comment]
> and then LGTM

I'll also spend the better part of tmorrow running the device tests for all the hardware in my office...

**Comment by superm1** [comment]
Just need to fixup the commit message for e4d9c29bef45273fdc92d9a9402ef9d8cf7211f4 to be one new GUID

**Comment by hughsie** [comment]
> Just

I'm about 2/3rds through testing all the hardware in my lab. Will fix up before pushing, thanks.

File: plugins/logitech-hidpp/fu-logitech-hidpp-peripheral.c [hunk 4]
```diff
@@ -611,6 +611,7 @@ fu_logitech_hidpp_peripheral_setup (FuDevice *device, GError **error)
 	if (idx != 0x00) {
 		self->is_updatable = TRUE;
 		fu_device_add_flag (FU_DEVICE (device), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+		fu_device_set_protocol (FU_DEVICE (device), "com.logitech.unifyingsigned");
```
Comment (superm1) [inline plugins/logitech-hidpp/fu-logitech-hidpp-peripheral.c:614]: this seems like an incorrect change - isn't the protocol signed only when it has idx `HIDPP_FEATURE_DFU_CONTROL_SIGNED` set?

Comment (superm1) [reply plugins/logitech-hidpp/fu-logitech-hidpp-peripheral.c:614]: and furthermore the protocol was actually originally set in `fu_logitech_hidpp_peripheral_init`

Comment (hughsie) [reply plugins/logitech-hidpp/fu-logitech-hidpp-peripheral.c:614]: Yes, but you only get `HIDPP_FEATURE_DFU_CONTROL` or `HIDPP_FEATURE_DFU_CONTROL_SIGNED` when you're in runtime mode. If you're stuck in recovery mode (e.g. you unplugged during update) you only get HIDPP_FEATURE_DFU which doesn't tell you what protocol we should be verifying. I guess the real fix would be to change `fu_device_set_protocol()` into `fu_device_add_protocol()` but that's a lot of work for something that I don't think is very common.

Comment (hughsie) [reply plugins/logitech-hidpp/fu-logitech-hidpp-peripheral.c:614]: I'll split this commit out as a new PR as it's more controversial it seems.

**Comment by thx1111** [comment]
Thanks Richard and Mario.  I tested this on Arch fwupd-git 1.5.7.r13.g0b6c67a3-1.  It recognized both receivers properly and also performed an upgrade properly on the correct second receiver.

**Comment by hughsie** [comment]
Woo, thanks for confirming! It was a pretty solid few days work, and I think I've tested every hit of hardware in my office validating the invasive change didn't break anything :)

**Pull request merged**